src/Microsoft.IdentityModel.JsonWebTokens/Json/JsonWebToken.HeaderClaimSet.cs

```diff
@@ -86,5 +53,65 @@ internal JsonClaimSet CreateHeaderClaimSet(ReadOnlySpan<byte> byteSpan)
+        /// Can be overridden to read and add custom claims.
+        /// If a custom claim is read, the reader should be positioned at the next token after reading the claim.
+        /// </summary>
+        /// <param name="reader">The Utf8JsonReader instance positioned at a claim name token used to read the JSON payload.</param>
+        /// <param name="claims">Collection of claims that have been read from the reader.</param>
+        private protected virtual void ReadHeaderValue(ref Utf8JsonReader reader, IDictionary<string, object> claims)
+        {
+            if (reader.ValueTextEquals(JwtHeaderUtf8Bytes.Alg))
+            {
+                claims[JwtHeaderParameterNames.Alg] = JsonSerializerPrimitives.ReadString(ref reader, JwtHeaderParameterNames.Alg, ClassName, true);
+            }
+            else if (reader.ValueTextEquals(JwtHeaderUtf8Bytes.Cty))
+            {
+                claims[JwtHeaderParameterNames.Cty] = JsonSerializerPrimitives.ReadString(ref reader, JwtHeaderParameterNames.Cty, ClassName, true);
+            }
+            else if (reader.ValueTextEquals(JwtHeaderUtf8Bytes.Kid))
+            {
+                claims[JwtHeaderParameterNames.Kid] = JsonSerializerPrimitives.ReadString(ref reader, JwtHeaderParameterNames.Kid, ClassName, true);
+            }
+            else if (reader.ValueTextEquals(JwtHeaderUtf8Bytes.Typ))
+            {
+                claims[JwtHeaderParameterNames.Typ] = JsonSerializerPrimitives.ReadString(ref reader, JwtHeaderParameterNames.Typ, ClassName, true); ;
+            }
+            else if (reader.ValueTextEquals(JwtHeaderUtf8Bytes.X5t))
+            {
+                claims[JwtHeaderParameterNames.X5t] = JsonSerializerPrimitives.ReadString(ref reader, JwtHeaderParameterNames.X5t, ClassName, true);
+            }
+            else if (reader.ValueTextEquals(JwtHeaderUtf8Bytes.Zip))
+            {
+                claims[JwtHeaderParameterNames.Zip] = JsonSerializerPrimitives.ReadString(ref reader, JwtHeaderParameterNames.Zip, ClassName, true);
+            }
+            else
+            {
+                string claimName = reader.GetString();
+
+                if (TryReadJwtClaim != null)
+                {
+                    reader.Read(); // Move to the value
+                    if (TryReadJwtClaim(ref reader, JwtSegmentType.Header, claimName, out object claimValue))
+                    {
+                        claims[claimName] = claimValue;
+                        reader.Read(); // Move to the next token
+                    }
+                    else
+                    {
+                        // The reader is positioned at the value token. The custom delegate did not read the value. Use our own logic.
+                        claims[claimName] = JsonSerializerPrimitives.ReadPropertyValueAsObject(ref reader, claimName, JsonClaimSet.ClassName, false);
```

Comment:
Can you comment on the class name here?

Reply:
This is from the existing code. It just ends up being used in logs: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/faa4998627289e752124c0f5ceb6e0c4c453bc18/src/Microsoft.IdentityModel.Tokens/Json/JsonSerializerPrimitives.cs#L41